src/annotator/test/frame-observer-test.js

```diff
@@ -4,140 +4,191 @@ import {
-    assert.calledWith(onFrameRemoved, frame);
-  });
+      assert.calledOnce(onFrameRemoved);
+      assert.calledWith(onFrameRemoved, iframe);
+    });
 
-  it('changing the `src` attribute triggers onFrameRemoved', async () => {
-    const frame = createAnnotatableIFrame();
-    await waitForFrameObserver();
+    it('changing the `src` attribute triggers onFrameRemoved', async () => {
+      const iframe = createAnnotatableIFrame();
+      await waitForFrameObserver();
 
-    assert.calledOnce(onFrameAdded);
-    assert.calledWith(onFrameAdded, frame);
+      assert.calledOnce(onFrameAdded);
+      assert.calledWith(onFrameAdded, iframe);
 
-    frame.setAttribute('src', document.location);
-    await waitForIFrameUnload(frame);
+      iframe.setAttribute('src', document.location);
+      await waitForIFrameUnload(iframe);
 
-    assert.calledOnce(onFrameRemoved);
-    assert.calledWith(onFrameRemoved, frame);
-  });
+      assert.calledOnce(onFrameRemoved);
+      assert.calledWith(onFrameRemoved, iframe);
+    });
+
+    it(`doesn't call onFrameAdded if FrameObserver is disconnected`, async () => {
+      frameObserver.disconnect();
+      const iframe = createAnnotatableIFrame();
+
+      frameObserver._discoverFrames(); // Emulate a race condition
+      await onDocumentReady(iframe);
 
-  it(`doesn't call onFrameAdded if FrameObserver is disconnected`, async () => {
-    frameObserver.disconnect();
-    const frame = createAnnotatableIFrame();
+      assert.notCalled(onFrameAdded);
+    });
 
-    frameObserver._discoverFrames(); // Emulate a race condition
-    await onDocumentReady(frame);
+    it("doesn't trigger onFrameAdded when annotatable iframe is from a different domain", async () => {
+      const iframe = createAnnotatableIFrame();
+      iframe.setAttribute('src', 'http://cross-origin.dummy');
 
-    assert.notCalled(onFrameAdded);
+      await onDocumentReady(iframe);
+      await waitForFrameObserver();
+
+      assert.notCalled(onFrameAdded);
+      assert.calledOnce(console.warn);
+    });
   });
 
-  // This test doesn't work. Surprisingly, `isAccessible` returns `true` even
-  // thought the iframe is from a different domain. My suspicion is that the
-  // iframe is accessed at a time where the loading has not yet started and it
-  // is therefore accessible.
-  // it("doesn't triggers onFrameAdded when annotatable iframe is from a different domain", async () => {
-  //   const frame = createAnnotatableIFrame('src', 'https://example.com');
-  //   frame.setAttribute('enable-annotation', '');
-  //   await waitForFrameObserver();
-
-  //   assert.notCalled(onFrameAdded);
-  // });
+  describe('onDocumentReady', () => {
+    let fakeIFrame;
+    let fakeIFrameDocument;
+
+    beforeEach(() => {
+      fakeIFrameDocument = {
+        addEventListener: sinon.stub(),
+        readyState: 'loading',
+        location: {
+          href: 'about:blank',
+        },
+      };
+
+      fakeIFrame = {
```

Comment:
I'd probably suggest to create a fake generic DOM element here and add the `contentWindow` property. Then you're only having to take the `contentWindow` part.

Reply:
I tried it in this way, but it didn't work:

```
      fakeIFrame = document.createElement('iframe');
      fakeIFrame.contentDocument = fakeIFrameDocument;  // TypeError: Cannot set property contentDocument of #<HTMLIFrameElement> which has only a getter
      fakeIFrame.setAttribute('src', 'http://my.dummy');
```

How do you suggest to create this fake generic DOM element?

Reply:
> How do you suggest to create this fake generic DOM element?

I meant using `document.createElement('div')` or some other general-purpose non-iframe element which doesn't have its own `contentDocument` and `contentWindow` accessors, so you can define custom ones.

Reply:
Yes, no problem, I can do that. Done.